org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/InstanceOfExpression.java

```diff
@@ -154,101 +149,64 @@ public void generateOptimizedBoolean(BlockScope currentScope, CodeStream codeStr
 		super.generateOptimizedBoolean(currentScope, codeStream, trueLabel, falseLabel, valueRequired);
 		return;
 	}
-	Constant cst = optimizedBooleanConstant();
 	addPatternVariables(currentScope, codeStream);
 
 	int pc = codeStream.position;
 
-	addAssignment(currentScope, codeStream, this.secretInstanceOfPatternExpressionValue);
-	codeStream.load(this.secretInstanceOfPatternExpressionValue);
+	this.expression.generateCode(currentScope, codeStream, true);
+	if (this.secretExpressionValue != null) {
+		codeStream.store(this.secretExpressionValue, true);
+		codeStream.addVariable(this.secretExpressionValue); // off by 1, but shouldn't matter
+	}
 
 	BranchLabel nextSibling = falseLabel != null ? falseLabel : new BranchLabel(codeStream);
 	codeStream.instance_of(this.type, this.type.resolvedType);
 	codeStream.ifeq(nextSibling);
-	codeStream.load(this.secretInstanceOfPatternExpressionValue);
+	if (this.secretExpressionValue != null) {
+		codeStream.load(this.secretExpressionValue);
+	} else {
+		this.expression.generateCode(currentScope, codeStream, true);
+	}
+
 	if (this.pattern instanceof RecordPattern) {
 		this.pattern.generateOptimizedBoolean(currentScope, codeStream, trueLabel, nextSibling);
-		codeStream.load(this.secretInstanceOfPatternExpressionValue);
-		codeStream.checkcast(this.type, this.type.resolvedType, codeStream.position);
 	} else {
 		codeStream.checkcast(this.type, this.type.resolvedType, codeStream.position);
-		codeStream.dup();
 		codeStream.store(this.elementVariable.binding, false);
 	}
 
-	codeStream.load(this.secretInstanceOfPatternExpressionValue);
-	codeStream.removeVariable(this.secretInstanceOfPatternExpressionValue);
-	codeStream.checkcast(this.type, this.type.resolvedType, codeStream.position);
+	if (this.secretExpressionValue != null) {
+		codeStream.removeVariable(this.secretExpressionValue);
+	}
 
-	if (valueRequired && cst == Constant.NotAConstant) {
+	if (valueRequired) {
 		codeStream.generateImplicitConversion(this.implicitConversion);
 	} else {
 		codeStream.pop();
 	}
 	codeStream.recordPositionsFrom(pc, this.sourceStart);
 
-
-	if ((cst != Constant.NotAConstant) && (cst.typeID() == TypeIds.T_boolean)) {
-		pc = codeStream.position;
-		if (cst.booleanValue() == true) {
-			// constant == true
-			if (valueRequired) {
-				if (falseLabel == null) {
-					// implicit falling through the FALSE case
-					if (trueLabel != null) {
-						codeStream.goto_(trueLabel);
-					}
-				}
```

Comment:
This block is dead code. See the first line in `org.eclipse.jdt.internal.compiler.ast.InstanceOfExpression.resolveType(BlockScope)` which does this:
```
this.constant = Constant.NotAConstant;
```